Source/WebCore/page/DragController.cpp

```diff
@@ -239,74 +239,99 @@ inline static bool dragIsHandledByDocument(DragHandlingMethod dragHandlingMethod
     if (RefPtr document = m_documentUnderMouse)
         shouldOpenExternalURLsPolicy = document->shouldOpenExternalURLsPolicyToPropagate();
 
-    if ((m_dragDestinationActionMask.contains(DragDestinationAction::DHTML)) && dragIsHandledByDocument(m_dragHandlingMethod)) {
-        client().willPerformDragDestinationAction(DragDestinationAction::DHTML, dragData);
-        bool preventedDefault = false;
-        if (localMainFrame->view())
-            preventedDefault = localMainFrame->eventHandler().performDragAndDrop(createMouseEvent(dragData), Pasteboard::create(dragData), dragData.draggingSourceOperationMask(), dragData.containsFiles());
+    auto continuePerformDrag = [weakThis = WeakPtr { *this }, this, completionHandler = WTFMove(completionHandler), dragData, weakFrame = WeakPtr { frame }, shouldOpenExternalURLsPolicy](bool preventedDefault) mutable {
+        if (!weakThis || !weakFrame) {
+            completionHandler(false);
+            return;
+        }
         if (preventedDefault) {
             clearDragCaret();
             m_documentUnderMouse = nullptr;
-            return true;
+            completionHandler(true);
+            return;
+        }
+        if ((m_dragDestinationActionMask.contains(DragDestinationAction::Edit)) && concludeEditDrag(dragData)) {
+            client().didConcludeEditDrag();
+            m_documentUnderMouse = nullptr;
+            clearDragCaret();
+            completionHandler(true);
+            return;
         }
-    }
 
-    if ((m_dragDestinationActionMask.contains(DragDestinationAction::Edit)) && concludeEditDrag(dragData)) {
-        client().didConcludeEditDrag();
         m_documentUnderMouse = nullptr;
         clearDragCaret();
-        return true;
-    }
-
-    m_documentUnderMouse = nullptr;
-    clearDragCaret();
-
-    if (!operationForLoad(dragData))
-        return false;
 
-    auto urlString = dragData.asURL();
-    if (urlString.isEmpty())
-        return false;
-
-    client().willPerformDragDestinationAction(DragDestinationAction::Load, dragData);
-    ResourceRequest resourceRequest { WTFMove(urlString) };
-    resourceRequest.setIsAppInitiated(false);
-    FrameLoadRequest frameLoadRequest { *localMainFrame, WTFMove(resourceRequest) };
-    frameLoadRequest.setShouldOpenExternalURLsPolicy(shouldOpenExternalURLsPolicy);
-    frameLoadRequest.setIsRequestFromClientOrUserInput();
-    localMainFrame->loader().load(WTFMove(frameLoadRequest));
-    return true;
+        if (!operationForLoad(dragData)) {
+            completionHandler(false);
+            return;
+        }
+        auto frameID = weakFrame->frameID();
+        client().willPerformDragDestinationAction(DragDestinationAction::Load, dragData, frameID, [weakThis = WTFMove(weakThis), completionHandler = WTFMove(completionHandler), weakFrame = WTFMove(weakFrame), shouldOpenExternalURLsPolicy, dragData]() mutable {
+            if (!weakThis)
+                return;
```

Comment:
```suggestion
                return completionHandler(false);
```

You probably still want to invoke the `completionHandler` here, if you go down this early return.